src/server/licensing_func.c

```diff
@@ -0,0 +1,971 @@
+ * terms of the GNU Affero General Public License as published by the Free
+ * Software Foundation, either version 3 of the License, or (at your option) any
+ * later version.
+ *
+ * PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Commercial License Information:
+ *
+ * For a copy of the commercial license terms and conditions,
+ * go to: (http://www.pbspro.com/UserArea/agreement.html)
+ * or contact the Altair Legal Department.
+ *
+ * Altair’s dual-license business model allows companies, individuals, and
+ * organizations to create proprietary derivative works of PBS Pro and
+ * distribute them - whether embedded or bundled with other software -
+ * under a commercial license agreement.
+ *
+ * Use of Altair’s trademarks, including but not limited to "PBS™",
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+/**
+ * @file    licensing_func.c
+ *
+ * @brief
+ * 		licensing_func.c - miscellaneous server functions
+ *
+ * Functions included are:
+ *
+ */
+
+#include "pbs_nodes.h"
+#include "pbs_license.h"
+#include "server.h"
+#include "work_task.h"
+#include "liblicense.h"
+#include "svrfunc.h"
+
+#define TEMP_BUF_LEN 20
+
+pbs_licensing_control licensing_control;
+pbs_license_counts license_counts;
+struct work_task *init_licensing_task;
+struct work_task *get_more_licenses_task;
+struct work_task *licenses_linger_time_task;
+extern void release_licenses(int);
+/**
+ * @brief
+ * 	consume_licenses - use count licenses from the pool of
+ * 			   already checked out licenses.
+ *
+ * @param[in]	- count	- number of licenses to consume
+ *
+ * @return int
+ * @retval  0 - success - able to consume count licenses
+ * @retval -1 - not enough licenses available.
+ */
+static int
+consume_licenses(long count)
+{
+	if (count <= license_counts.licenses_local) {
+		license_counts.licenses_local -= count;
+		license_counts.licenses_used += count;
+		return 0;
+	} else
+		return -1;
+}
+
+/**
+ * @brief
+ * 	return_licenses - return count licenses back to the
+ * 			  pool of already checked out licenses.
+ *
+ * @param[in]	- count	- number of licenses to consume
+ *
+ * @return void
+ */
+static void
+return_licenses(long count)
+{
+	license_counts.licenses_local += count;
+	license_counts.licenses_used -= count;
+}
+
+/**
+ * @brief
+ * 	add_to_unlicensed_node_list - add a node to the list
+ * 				      of unlicensed nodes
+ *
+ * @param[in]	- index - index of the node
+ *
+ * @return void
+ */
+static void
+add_to_unlicensed_node_list(int index)
```

Comment:
we are storing the index in the list - how would passing pnode help?

Reply:
You cannot rely on the index as it can be changed when you delete some node.

Reply:
When some node will be deleted again we will have sorted index and whenever the function will be called it will be getting the updated index of the particular node and execution of the function will be performed according to that.

Reply:
Agree on that with you @nithinj, making the change to pass pnode